### PR TITLE
fix(api): stop intermittent 503s and cut progress read bandwidth

### PR DESCRIPTION
**Pull request by DysektAI**
## Why

`GET /api/v2/progress` and `/team/progress` intermittently returned **503 "Rate limiter unavailable"** to real authenticated users (~10/day, wall times clustering right at the client timeout). Issue #574 traced this to the shape of the rate-limit check, not to load or CPU limits.

The gateway checked its three buckets — daily quota, per-minute burst, per-IP backstop — **sequentially**, each with its own 3s `AbortController` budget against the `ApiGatewayRateLimiter` Durable Object. Worst case was 9s of gateway time, and because the primary buckets fail closed by design, a slow DO turned into a user-visible 503. Only the timeout path logged anything, so the 503s were never attributable to a specific bucket or cause.

While in here, this PR also lands the bandwidth work from #524 / #361, since the same request path was fetching roughly twice the data it needed and disabling every form of conditional request.

## Rate limiting (fixes the 503s)

- **Parallel bucket checks.** Daily, burst, and IP now run under one `Promise.all`. Worst case is a single timeout budget instead of three sequential ones.
- **Refund reconciliation.** Because all three buckets are consulted concurrently, a bucket can consume a slot for a request that another bucket then denies. Every bucket that allowed is refunded whenever the request is not served, preserving the previous invariant that *a denied request drains no quota* — including on the fail-closed 503 path, which previously leaked slots.
- **Timeout 3s → 5s, plus `Retry-After` on the 503.** 3s was aggressive for a SQLite-backed DO under burst load. Clients now get an explicit backoff hint instead of hot-looping.
- **Structured `rate_limiter_unavailable` logs on all four unavailability paths** (`http_status`, `bad_json`, `timeout`, `fetch_error`) with `action`, `reason`, `duration_ms`, `timeout_ms`. Infrastructure failures are never logged as `rate_limit_429` and never recorded as throttles in `api_usage_daily`, so the two failure classes stop being conflated in dashboards.
- **Per-IP keys no longer opt into retention.** `rateLimit()` unconditionally sent `retain: true`, including for high-cardinality `ip-*` keys, so every distinct client IP left DO storage behind forever. IP keys now schedule DO self-cleanup; user-keyed daily/burst buckets still retain.

## Progress reads (bandwidth)

- **Narrowed the Supabase select.** `select=*` was returning both `pvp_data` and `pve_data` on every read; now only the token's game-mode column is fetched. Applies to `/progress`, `/team/progress`, and the solo fallback.
- **Conditional reads.** Weak `ETag` + `If-None-Match` → `304 Not Modified`, `Cache-Control: private, max-age=15`, `Vary: Accept-Encoding, Authorization, Origin`.
- **gzip** for bodies ≥1 KiB when the client sends `Accept-Encoding: gzip`.

One deliberate design note: the ETag is derived from the **serialized response payload**, not from `user_progress.updated_at` as #524 originally proposed. Progress responses fold in task metadata from `json.tarkov.dev`, invalidation results, and game-edition hideout auto-completes — none of which touch `updated_at`. An `updated_at`-based validator would return `304` while the real payload had changed. Hashing the payload costs one SHA-256 over an already-serialized string and cannot go stale.

## Debuggability

The recurring complaint in #574 was not knowing which layer a failure came from. Concretely:

- A `503` now always carries a `rate_limiter_unavailable` log line naming the bucket and reason.
- A `429` carries `rate_limit_429` with `bucket` ∈ `daily | burst | ip`.
- Infrastructure failures and quota denials no longer share a code path, a log shape, or a usage row.
- `docs/SYSTEMS.md` gains a **Progress API data flow** section: the full hop map from client to response, the implementing files, and seven invariants an agent or reviewer can check against the code.

## Docs

- `docs/SYSTEMS.md` — new section 6 with diagram, flow, files, invariants.
- `docs/RATE_LIMITING.md` — parallel-check diagram, refund semantics, retention split, updated fail-behavior table.
- `docs/API.md` — new "Conditional Requests & Polling" section; 503-is-not-a-quota-violation note.
- `workers/api-gateway/src/openapi.ts` — `2.3.0` → `2.4.0`, `If-None-Match` parameter, `304` response, ≥60s polling guidance.

## Validation

- `pnpm run test:api-gateway` — 128 passed (10 new: parallel dispatch, fail-closed 503 + refunds, daily-denies-refunds-others, IP-key retention, narrowed select, ETag/304 hit and miss, gzip on/off/below-threshold).
- `pnpm run typecheck` — pass; `tsc -p workers/api-gateway` — pass.
- `pnpm run lint` — pass. `eslint workers/api-gateway/src` reports 35 warnings, byte-identical to `main` and all in files this PR does not touch.
- `pnpm run validate:openapi` — pass.
- `node scripts/check-systems-drift.mjs` — pass.

## Issue disposition

- **#574** — P1 and P2 acceptance criteria are met, plus P4 docs. The P3 architecture decision (native Rate Limiting binding vs. single batched DO call vs. KV snapshots) is intentionally **not** included; it needs a `docs/decisions/` ADR and should be judged after this change's effect on the 503 rate is observable in production. Suggest leaving #574 open for P3 only.
- **#524** — fully addressed (ETag/`If-None-Match`/304, `Cache-Control`, polling docs). Ready to close on merge.
- **#361** — the remaining conditional-reads and polling-documentation items are done here. The Logpush/WAF work stays with #525.
- **#529** — **recommend closing as superseded.** That issue proposed deleting the `alarm()` handler once pre-deployment alarms had drained. Making the per-IP keys ephemeral (above) makes `alarm()` permanently load-bearing: it is now the mechanism that reclaims high-cardinality IP-key storage. Only the non-ephemeral *transitional drain branch* inside `alarm()` remains disposable, and it is cheap to keep.

## Not included

- The #574 P3 architecture change (needs an ADR and post-deploy data).
- Cloudflare Logpush / WAF rules (#525) — infra configuration, not code.

---

<a href="https://app.devin.ai/review/tarkovtracker-org/tarkovtracker/pull/589" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

<a href="https://cubic.dev/pr/tarkovtracker-org/TarkovTracker/pull/589?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

___

## **CodeAnt-AI Description**
**Stop intermittent rate-limiter failures and reduce progress read bandwidth**

### What Changed
- `GET /progress` and `GET /team/progress` now return a weak `ETag`, support `If-None-Match`, and send `304 Not Modified` when the payload has not changed.
- Progress responses now include short private caching and gzip compression for larger responses when the client asks for it.
- Progress reads fetch only the active game mode’s data instead of both modes, cutting the amount of data returned from storage.
- Rate-limit checks now complete in parallel, which removes the long waits that were causing occasional `503 Rate limiter unavailable` errors.
- When a request is rejected, any quota already consumed by other checks is refunded; primary limiter failures now return `503` with `Retry-After: 10`, while IP backstop failures still allow the request through.

### Impact
`✅ Fewer intermittent 503s on progress reads`
`✅ Lower bandwidth for progress polling`
`✅ Smaller responses for cached progress requests`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

## Summary by CodeRabbit

* **New Features**
  * Added conditional requests for progress endpoints, including ETag validation and `304 Not Modified` responses.
  * Added private caching headers and negotiated gzip compression for larger responses.
  * Progress data retrieval now returns only the selected game mode’s data.

* **Bug Fixes**
  * Improved rate-limit handling, quota refunds, and unavailable-limiter responses with retry guidance.
  * Refined IP backstop behavior so infrastructure failures do not unnecessarily block requests.

* **Documentation**
  * Updated API, rate-limiting, systems, and OpenAPI documentation with the new behavior.

<details open><summary><h3>Greptile Summary</h3></summary>

This PR fixes intermittent 503s on `GET /progress` and `/team/progress` by parallelizing the three rate-limit bucket checks, adds proper refund reconciliation for denied requests, narrows Supabase selects to the active game-mode column only, and adds ETag/304/gzip support for progress endpoints.

- **Parallel rate limiting**: daily, burst, and IP checks now run under a single `Promise.all`; any bucket that consumed a slot is refunded via `ctx.waitUntil` when another bucket denies or the primary limiter is unavailable, preserving the "a denied request drains no quota" invariant. IP keys no longer carry `retain: true`, so high-cardinality IP Durable Objects self-clean via alarms.
- **Bandwidth**: `select=*` is replaced with `select=user_id,game_edition,${dataField}`, fetching only the requested mode's JSONB blob; progress responses now carry a payload-derived weak ETag, `If-None-Match → 304`, `Cache-Control: private, max-age=15`, and optional gzip for bodies ≥ 1 KiB. The new `UserProgressModeRow` projected type keeps the compiler honest about which columns are absent.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge — the parallel rate-limit rewrite is internally consistent, refund semantics are correct, and ETag/gzip logic is RFC-compliant.

The three bucket checks now run concurrently under a single timeout budget, eliminating the worst-case sequential stall that caused 503s. Refund reconciliation correctly handles every denial path — including the fail-closed 503 path that previously leaked slots. The ETag is derived from the same serialized bytes sent to the client, so the validator and the body cannot disagree. The narrowed Supabase select is matched by a projected type that prevents compiler blind-spots. Test coverage is thorough, including the parallel-dispatch deadlock test, 503+refund invariant, and parameterized Accept-Encoding edge cases.

No files require special attention.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| workers/api-gateway/src/index.ts | Core logic change: rate-limit checks parallelized with correct refund semantics; ETag/304/gzip conditional-read infrastructure added; IP key retention removed; structured logging for unavailability paths. |
| workers/api-gateway/src/handlers/progress.ts | Narrowed Supabase select to user_id,game_edition,{mode}_data; switched to UserProgressModeRow projected type; no logic issues found. |
| workers/api-gateway/src/handlers/team.ts | Same narrowed select applied consistently to team and solo fallback paths; type updated to UserProgressModeRow. |
| workers/api-gateway/src/types.ts | Adds UserProgressModeRow as a Pick+Partial projection of UserProgressRow, correctly reflecting that only one JSONB column is fetched per request. |
| workers/api-gateway/src/__tests__/gateway.test.ts | Adds 10+ tests covering narrowed select, ETag/304 round-trip, gzip on/off/below-threshold, and parameterized Accept-Encoding edge cases including q=0 rejection. |
| workers/api-gateway/src/__tests__/rate-limits.test.ts | Adds tests for parallel dispatch, fail-closed 503 + refunds, daily-denies-refunds-others, and IP-key no-retention invariant. |
| workers/api-gateway/src/openapi.ts | Version bump to 2.4.0, adds If-None-Match parameter and 304 response, and polling guidance; straightforward documentation update. |

</details>

<details open><summary><h3>Sequence Diagram</h3></summary>

```mermaid
sequenceDiagram
    participant Client
    participant Worker as API Gateway Worker
    participant DO as ApiGatewayRateLimiter DO
    participant Supabase

    Client->>Worker: GET /progress
    Worker->>Worker: Validate token

    par Parallel rate-limit checks
        Worker->>DO: "POST daily-read:{user_id} (retain:true)"
        Worker->>DO: "POST burst-read:{user_id} (retain:true)"
        Worker->>DO: "POST ip-read:{ip} (no retain)"
    end

    alt Primary limiter 503
        Worker->>DO: refund consumed slots (waitUntil)
        Worker-->>Client: 503 + Retry-After: 10
    else Daily/burst deny 429
        Worker->>DO: refund other slots (waitUntil)
        Worker-->>Client: "429 + X-RateLimit-*"
    else IP backstop 429
        Worker->>DO: refund daily + burst (waitUntil)
        Worker-->>Client: 429
    else IP backstop 503
        Note over Worker: fail open, log warning
    end

    Worker->>Supabase: "GET /rest/v1/user_progress?select=user_id,game_edition,pvp_data"
    Worker->>Worker: transformProgress() → JSON → SHA-256 ETag

    alt If-None-Match matches
        Worker-->>Client: 304 + ETag + Cache-Control
    else Large body + gzip accepted
        Worker-->>Client: 200 + Content-Encoding: gzip + ETag
    else
        Worker-->>Client: 200 + ETag + Cache-Control
    end
```
</details>

<sub>Reviews (4): Last reviewed commit: ["docs(api): declare Vary and numeric boun..."](https://github.com/tarkovtracker-org/tarkovtracker/commit/5e611697bd4aec230fd4ca976fd50f50845cb95c) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=46706771)</sub>

**Context used:**

- Context used - AGENTS.md ([source](https://app.greptile.com/dysektai/github/tarkovtracker-org/tarkovtracker/-/custom-context?memory=635c24fb-71e5-43a4-9979-74405b725e95))

### CODE REVIEW

File: docs/SYSTEMS.md [hunk 1]
```diff
@@ -397,6 +398,97 @@ flowchart LR
...
+4. **Data fetch.** `workers/api-gateway/src/handlers/progress.ts` selects only
+   `user_id,game_edition,<mode>_data` from `user_progress` (never both game modes), resolves a
+   display-name fallback from Supabase Auth (24h cache), and loads tasks/hideout metadata from
+   `json.tarkov.dev` via `workers/api-gateway/src/services/tarkov.ts` (1h memory cache).
+5. **Transform.** `workers/api-gateway/src/utils/transform.ts` converts the JSONB objects into the
+   public array format, applies invalidation (`workers/api-gateway/src/utils/invalidation.ts`) and
+   game-edition hideout auto-completes.
+6. **Response.** `conditionalReadResponse` in `workers/api-gateway/src/index.ts` serializes once,
+   derives a weak `ETag` from the payload, answers `304` on a matching `If-None-Match`, sets
+   `Cache-Control: private, max-age=15` + `Vary: Accept-Encoding, Authorization`, and gzips bodies
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply docs/SYSTEMS.md]: Valid — fixed in be7b5822. The `Origin` value was added to `READ_VARY` late in review after I noticed the CORS allow-origin header is echoed per request, and I did not carry it back into the doc. Step 6 of the SYSTEMS.md flow now reads `Vary: Accept-Encoding, Authorization, Origin` and also records the `gzip;q=0` behavior. `node scripts/check-systems-drift.mjs` passes.

File: workers/api-gateway/src/index.ts [hunk 2]
```diff
@@ -607,6 +623,66 @@ function errorResponse(
...
+    Vary: READ_VARY,
+    ETag: etag,
+    ...(extraHeaders ?? {}),
+  };
+  if (etagMatches(request.headers.get('If-None-Match'), etag)) {
+    return new Response(null, { status: 304, headers });
+  }
+  headers['Content-Type'] = 'application/json';
+  const acceptEncoding = request.headers.get('Accept-Encoding') || '';
+  if (/(^|[\s,])gzip($|[\s,;])/i.test(acceptEncoding) && json.length >= GZIP_MIN_BYTES) {
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply workers/api-gateway/src/index.ts]: Valid, and thanks for catching it — fixed in be7b5822. `gzip;q=0` is a rejection under RFC 9110 §12.5.3 and the old regex accepted it because `;` was in the trailing class.

I went with a small parser rather than the suggested regex, because the regex still can't express the `*` precedence rule and is hard to audit:

```ts
function acceptsGzip(header: string | null): boolean {
  if (!header) return false;
  let wildcard = false;
  for (const part of header.split(',')) {
    const [rawCoding, ...params] = part.split(';');
    const coding = rawCoding.trim().toLowerCase();
    if (coding !== 'gzip' && coding !== '*') continue;
    const qParam = params.map((p) => p.trim().toLowerCase()).find((p) => p.startsWith('q='));
    const acceptable = !qParam || Number(qParam.slice(2)) > 0;
    if (coding === 'gzip') return acceptable;
    wildcard = acceptable;
  }
  return wildcard;
}
```

An explicit `gzip` entry wins over `*` regardless of order, and a malformed q-value falls back to uncompressed since that is always decodable. Covered by an 8-case table-driven test including `gzip;q=0`, `gzip;q=0.0`, `br, gzip;q=0`, `*;q=0, gzip`, and `gzip;q=0, *`.

File: workers/api-gateway/src/index.ts [hunk 3]
```diff
@@ -607,6 +623,66 @@ function errorResponse(
...
+  requestOrigin?: string,
+  extraHeaders?: Record<string, string>
+): Promise<Response> {
+  const body: Record<string, unknown> = { success: true };
+  if (data && typeof data === 'object' && 'data' in data) {
+    body.data = (data as Record<string, unknown>).data;
+    body.meta = (data as Record<string, unknown>).meta;
+  } else {
+    body.data = data;
+  }
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply workers/api-gateway/src/index.ts:689]: Fair point on the latent footgun — addressed in be7b5822. The envelope now only assigns `meta` when the handler actually supplied it, so there is no `undefined` value relying on `JSON.stringify` to drop it.

Worth noting the ETag could not have diverged from the body even before this change, since both derive from the same `JSON.stringify` call. That is now structurally enforced rather than incidental: the payload is encoded to bytes once and those same bytes back the digest, the gzip size threshold, and the response body.

File: workers/api-gateway/src/index.ts [hunk 4]
```diff
@@ -607,6 +623,66 @@ function errorResponse(
...
+  if (etagMatches(request.headers.get('If-None-Match'), etag)) {
+    return new Response(null, { status: 304, headers });
+  }
+  headers['Content-Type'] = 'application/json';
+  const acceptEncoding = request.headers.get('Accept-Encoding') || '';
+  if (/(^|[\s,])gzip($|[\s,;])/i.test(acceptEncoding) && json.length >= GZIP_MIN_BYTES) {
+    headers['Content-Encoding'] = 'gzip';
+    const stream = new Blob([json]).stream().pipeThrough(new CompressionStream('gzip'));
+    return new Response(stream, { status: 200, headers });
+  }
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply workers/api-gateway/src/index.ts]: Both noted, and the second one was a real inconsistency worth fixing — addressed in be7b5822.

The q-value handling is now RFC-correct (see the reply on the Greptile thread for the parser). On the size gate: you are right that `json.length` counted UTF-16 code units while the ETag digest hashed UTF-8 bytes. Rather than patch the comparison, the payload is now encoded to a `Uint8Array` once and the same bytes back the digest, the `byteLength >= GZIP_MIN_BYTES` check, and the response body — so the two can no longer disagree by construction. Display names are user-supplied and can be multi-byte, so this was reachable, if only right at the threshold.

File: workers/api-gateway/src/index.ts [hunk 5]
```diff
@@ -686,41 +762,71 @@ async function authenticateAndRateLimit(
...
+  // Availability policy: daily and burst are the primary limiters and fail
+  // closed — with them unavailable there is no protection, so the request is
+  // rejected with an honest 503 (not counted or logged as a throttle). The IP
+  // backstop is a secondary signal and fails open below.
+  if (daily.status === 503 || burst.status === 503) {
+    refundConsumed('unavailable');
+    return errorResponse('Rate limiter unavailable', 503, envOrigin, requestOrigin, {
+      'Retry-After': String(RATE_LIMIT_UNAVAILABLE_RETRY_AFTER_S),
+    });
+  }
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply workers/api-gateway/src/index.ts:848]: Thanks for the path-by-path audit — that matches the intended semantics, and the tradeoff you identify is the right one to name.

To be explicit about it: refunds are best-effort via `ctx.waitUntil`, so a failed refund can transiently leak a slot. That is strictly better than the behavior it replaces (the old sequential fail-closed 503 path refunded nothing at all), and the exposure is bounded — a leaked daily slot is 1/1000 of a free-tier day and self-heals at the 00:00 UTC reset; burst and IP slots age out of their sliding windows within 60s and 1h respectively. Recorded as an invariant in the new `docs/SYSTEMS.md` section 6.

File: workers/api-gateway/src/index.ts [hunk 6]
```diff
@@ -607,6 +623,66 @@ function errorResponse(
...
+  headers['Content-Type'] = 'application/json';
+  const acceptEncoding = request.headers.get('Accept-Encoding') || '';
+  if (/(^|[\s,])gzip($|[\s,;])/i.test(acceptEncoding) && json.length >= GZIP_MIN_BYTES) {
+    headers['Content-Encoding'] = 'gzip';
+    const stream = new Blob([json]).stream().pipeThrough(new CompressionStream('gzip'));
+    return new Response(stream, { status: 200, headers });
+  }
+  return new Response(json, { status: 200, headers });
+}
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply workers/api-gateway/src/index.ts:714]: Confirmed, and thanks for verifying the determinism explicitly — that is exactly the property the payload-derived validator depends on.

Worth adding for future readers: `crypto.randomUUID()` does appear in `handlers/progress.ts` via `buildApiUpdateMeta`, but only on the write path where it populates `lastApiUpdate`. That field is stored in the JSONB blob and is not projected into `ProgressResponseData` by `transformProgress`, so it never reaches a read payload or the digest.

File: workers/api-gateway/src/handlers/progress.ts [hunk 7]
```diff
@@ -285,8 +285,10 @@ export async function handleGetProgress(
   token: ApiToken,
   gameMode: 'pvp' | 'pve'
 ): Promise<ProgressResponse> {
-  // Fetch user progress from Supabase
-  const url = `${env.SUPABASE_URL}/rest/v1/user_progress?user_id=eq.${token.user_id}&select=*&limit=1`;
+  // Select only the requested game mode's JSONB blob — fetching both pvp_data
+  // and pve_data doubles Supabase egress and Worker memory for no benefit.
+  const dataField = gameMode === 'pve' ? 'pve_data' : 'pvp_data';
+  const url = `${env.SUPABASE_URL}/rest/v1/user_progress?user_id=eq.${token.user_id}&select=user_id,game_edition,${dataField}&limit=1`;
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply workers/api-gateway/src/handlers/progress.ts:291]: Confirmed — that matches my own check of the consumers. `updated_at` is the one field worth calling out: it remains on the `UserProgressRow` type and is now no longer selected, which is intentional. #524 originally proposed deriving the ETag from it, and this PR deliberately does not, since progress responses fold in `json.tarkov.dev` task metadata, invalidation results, and game-edition hideout auto-completes — none of which touch `updated_at`. A validator based on it would return `304` while the real payload had changed.

**Comment by DysektAI** [comment]
Addressed all 7 findings from the cubic ultrareview in 234057ef. Summary of dispositions:

**`handlers/progress.ts` — projection vs. `UserProgressRow`** — valid, fixed. This was the most useful catch: narrowing the select without narrowing the type meant `current_game_mode` and the unselected mode blob type-checked as present and would have returned `undefined` at runtime for a future reader. Added `UserProgressModeRow` (`Pick` of the selected columns + `Partial` of the two mode blobs) and widened `extractGameModeData` to accept it. `UserProgressRow` remains assignable, so nothing else had to change.

**`index.ts` — out-of-range qvalues** — valid, fixed. `Number(q) > 0` read `q=2` as a stronger preference rather than as malformed, contradicting the function's own comment. qvalues are now validated against the RFC 9110 grammar (`0[.0-3 digits]` / `1[.up to three zeros]`) before the range check. Test table extended to 11 cases covering `q=2`, `q=abc`, and `q=1.0`.

**`openapi.ts` — undeclared `Retry-After` on 503 and no headers on 304** — valid, fixed. Both were prose-only contracts that generated clients could not see. `ServiceUnavailable` now declares `Retry-After`; `NotModified` declares `ETag`, `Cache-Control`, and the three `X-RateLimit-*` headers. Declaring `ETag` on the 304 matters most, since a polling client needs it to build the next conditional request.

**`docs/API.md` — over-broad fail-closed wording** — valid, fixed. The text said the gateway fails closed whenever "the rate limiter" is unavailable; that holds only for the per-user daily and burst buckets. Now states explicitly that the IP backstop fails open and that no `503` is ever attributable to it.

**`docs/API.md` — undocumented gzip threshold** — valid, fixed. Documented the 1 KiB floor and warned against assuming `Content-Encoding` is always present.

**`docs/SYSTEMS.md` — refund invariant overstated** — valid, fixed, and the sharpest of the doc findings. The invariant claimed a rejected request never leaves quota consumed, but refunds are best-effort `ctx.waitUntil` calls. Rewrote it to guarantee what the code actually guarantees — that a refund is *attempted* for every bucket that allowed — and to record why the residue is bounded and self-healing (daily reclaimed at the 00:00 UTC reset, burst/IP aged out of their sliding windows). Making this transactional would require durable retry semantics in the DO, which is not worth it at this failure rate; if that changes it belongs with the #574 P3 architecture decision.

Validation on 234057ef: `test:api-gateway` 139 passed, `tsc -p workers/api-gateway` pass, `nuxt typecheck` pass, `validate:openapi` pass, `check-systems-drift` pass, `eslint workers/api-gateway/src` unchanged from main (35 pre-existing warnings, none in touched files).

**Comment by DysektAI** [comment]
Addressed the three OpenAPI contract findings in 5e611697.

All three were valid and are about the spec under-describing what the gateway actually sends:

- **304 missing `Vary`** — the response declared `Cache-Control` but not `Vary: Accept-Encoding, Authorization, Origin`. That is the more consequential of the three, since a proxy or client that keys a cached 304 without the variant dimensions can serve a gzipped body to a client that did not ask for one, or cross token scopes. Now declared.
- **Unbounded 304 rate-limit header schemas** — `X-RateLimit-*` admitted negative limits and counts. Now reuses the exact minimums the existing 429 contract declares (`Limit` min 1, `Remaining`/`Reset` min 0), so the two contracts describing the same headers no longer disagree.
- **Unbounded 503 `Retry-After`** — permitted zero or negative delays on a response whose entire purpose is to instruct a wait. Now `minimum: 1`, matching the 429 `Retry-After`.

No runtime behavior changed; these are spec-only corrections. `validate:openapi` passes, `test:api-gateway` 139 passed, `tsc -p workers/api-gateway` passes.

**Comment by DysektAI** [comment]
## Disposition: superseded — closing in favor of #607

Reviewing this PR against current `main`: three follow-on PRs landed the bulk of its content independently, and the remaining piece (conditional reads) is being re-landed in **#607**. A straight rebase of this branch would require dropping most of the diff, so I'm closing it rather than rebasing.

### What landed independently

| #589 piece | Status | Landed via |
|---|---|---|
| Parallel 3-bucket rate-limit checks + refund reconciliation | **Superseded** — `main` now uses a single daily DO call + native CF `[[ratelimits]]` abuse gate (#591). The parallel-check/refund/IP-backstop approach here is a dead design. | #591 (merged 2026-07-24) |
| IP-key `retain: true` removal + DO self-cleanup | **Superseded** — #591 deleted the IP bucket entirely. | #591 |
| 503 `Retry-After`, structured `rate_limiter_unavailable` logs | **Superseded** — #591 fails open on DO unavailability (`daily_quota_unavailable`); #596 hardened abuse-gate logging and removed obsolete 503s from the OpenAPI contract. | #591 + #596 (merged 2026-07-29) |
| Narrowed Supabase select + `UserProgressModeRow` projection | **Already on `main`** | #590 (merged 2026-07-29) |
| OpenAPI 503 cleanup | **Superseded** | #596 |

### What was NOT on `main` and is being re-landed

The conditional-reads layer — the only piece of #589 not yet on `main` — is re-implemented in **#607**, rebased on current `main`:

- Weak `ETag` (payload-derived SHA-256) + `If-None-Match` → `304`
- `Cache-Control: private, max-age=15` + `Vary: Accept-Encoding, Authorization, Origin`
- RFC 9110 gzip negotiation for bodies ≥1 KiB (honoring `gzip;q=0`)
- OpenAPI `2.3.0` → `2.4.0`, `IfNoneMatchHeader` param, `NotModified` response, `304` on both read endpoints
- `docs/API.md` Conditional Requests & Polling section
- `docs/SYSTEMS.md` §6 Progress API data flow (adapted to the current single-DO + abuse-gate architecture)
- The ETag/304/gzip/Accept-Encoding tests from this PR

#607 carries forward the review-hardened logic from this PR's final commit (the gzip;q=0 fix, byte-length threshold, `Vary` on 304, payload-derived ETag rationale). The review feedback from cubic/Greptile/CodeRabbit on the conditional-read layer was all still valid and applies cleanly.

### Issue tracker

- **#524** — closed by #607.
- **#361** — conditional-reads portion done in #607; Logpush/WAF stays with #525.
- **#574** — P1/P2/P4 met by #591+#596; P3 (architecture ADR) intentionally deferred.
- **#529** — re-triage separately; #591's removal of the IP bucket changed the calculus.

Thanks for the thorough review pass on this — the findings directly shaped #607.

**Comment by DysektAI** [comment]
Closing as superseded — see the disposition comment above. The salvageable conditional-reads layer is re-landed in #607.
